packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,416 @@
+#ifndef _ZOLTAN2_DISTANCE2_HPP_
+#define _ZOLTAN2_DISTANCE2_HPP_
+
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <queue>
+#include <sys/time.h>
+
+#include "Zoltan2_Algorithm.hpp"
+#include "Zoltan2_GraphModel.hpp"
+#include "Zoltan2_ColoringSolution.hpp"
+#include "Zoltan2_Util.hpp"
+#include "Zoltan2_TPLTraits.hpp"
+#include "Zoltan2_AlltoAll.hpp"
+
+
+#include "Tpetra_Core.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Tpetra_Import.hpp"
+#include "Tpetra_FEMultiVector.hpp"
+
+#include "Kokkos_Core.hpp"
+#include "KokkosSparse_CrsMatrix.hpp"
+#include "KokkosKernels_Handle.hpp"
+#include "KokkosKernels_IOUtils.hpp"
+#include "KokkosGraph_Distance2Color.hpp"
+#include "KokkosGraph_Distance2ColorHandle.hpp"
+
+/////////////////////////////////////////////////
+//! \file Zoltan2_Distance1_2GhostLayer.hpp
+//! \brief A hybrid Distance-2 Coloring Algorithm
+
+
+namespace Zoltan2{
+
+template <typename Adapter>
+class AlgDistance2 : public AlgTwoGhostLayer<Adapter> {
+
+  public:
+    
+    using lno_t = typename Adapter::lno_t;
+    using gno_t = typename Adapter::gno_t;
+    using offset_t = typename Adapter::offset_t;
+    using scalar_t = typename Adapter::scalar_t;
+    using base_adapter_t = typename Adapter::base_adapter_t;
+    using map_t = Tpetra::Map<lno_t,gno_t>;
+    using femv_scalar_t = int;
+    using femv_t = Tpetra::FEMultiVector<femv_scalar_t, lno_t, gno_t>; 
+    using device_type = Tpetra::Map<>::device_type;//Kokkos::Device<Kokkos::Cuda, Kokkos::Cuda::memory_space>;
+    using execution_space = Tpetra::Map<>::execution_space;//Kokkos::Cuda;
+    using memory_space = Tpetra::Map<>::memory_space;//Kokkos::Cuda::memory_space;
+  private:
+    
+    virtual void colorInterior(const size_t nVtx,
+                       Kokkos::View<lno_t*,device_type> adjs_view,
+                       Kokkos::View<offset_t*, device_type> offset_view,
+                       Teuchos::RCP<femv_t> femv,
+		       Kokkos::View<lno_t*, device_type> vertex_list,
+		       size_t vertex_list_size = 0,
+                       bool recolor=false){
+      using KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle
+          <offset_t, lno_t, lno_t, execution_space, memory_space, memory_space>;
+  
+      KernelHandle kh;
+      kh.set_verbose(this->verbose); 
+      if(recolor)kh.create_distance2_graph_coloring_handle(KokkosGraph::COLORING_D2_VB_BIT);
+      else kh.create_distance2_graph_coloring_handle(KokkosGraph::COLORING_D2_NB_BIT);
+      if(vertex_list_size != 0){
+        kh.get_distance2_graph_coloring_handle()->set_vertex_list(vertex_list, vertex_list_size);
+      }
+      Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<memory_space>();
```

Comment:
If you know that you want a device view here, use getLocalViewDevice instead of templated version.
Also, type of femvColors may not be correct without UVM;  better to use "auto"